### ISSUE TITLE
Error 500 creating a PR

### ISSUE THREAD

**Issue by karthanistyr** [go-gitea/gitea#16936]
- Gitea version (or commit ref): v1.15.1 (docker image from docker hub)
- Git version: N/A
- Operating system: debian 10
  
  
  
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
- Log :
```
2021/09/02 22:52:16 ...ervices/pull/pull.go:39:NewPullRequest() [I] [SQL] BEGIN TRANSACTION [] - 355.924µs
2021/09/02 22:52:16 models/index.go:92:getNextResourceIndex() [I] [SQL] SELECT max_index FROM issue_index WHERE
group_id = $1 [1] - 485.227µs
2021/09/02 22:52:16 models/index.go:66:GetNextResourceIndex() [I] [SQL] INSERT INTO issue_index (group_id, max_index) VALUES ($1,1) ON CONFLICT (group_id) DO UPDATE SET max_index = issue_index.max_index+1 [1] - 551.813µs
2021/09/02 22:52:16 models/pull.go:431:NewPullRequest() [I] [SQL] ROLLBACK [] - 107.085µs
2021/09/02 22:52:16 ...ers/web/repo/pull.go:1113:CompareAndPullRequestPost() [E] NewPullRequest: generate issue
index failed: pq: there is no unique or exclusion constraint matching the ON CONFLICT specification
```

## Description

1. Created a branch locally
```
git checkout -b upgrade/mybranch
```
2. Pushed branch to gitea
```
git push origin upgrade/mybranch
```
3. On Gitea web app: create PR; error 500 occurs when committing the PR (clicking "Create Pull Request" button)

On a gitea running in docker; 1.15.0 used to work fine, but v1.15.1 does not.

**Comment by karthanistyr** [comment]
Note on my upgrade process (as per https://docs.gitea.io/en-us/install-with-docker/#upgrading)
```
docker-compose down
# gitea stop...
# upgrading the gitea version number in compose file
docker-compose pull
docker-compose up -d
```

**Comment by karthanistyr** [comment]
Maybe it looks like a db migration issue, as the specific migration script (models/migrations/v182.go) was amended, but my db may have had run it already, and so maybe is missing stuff? Just speculating here...

**Comment by karthanistyr** [comment]
I rolled back to 1.15.0 in the meantime.

**Comment by karthanistyr** [comment]
Flagging #16820 as potential culprit

**Comment by zeripath** [comment]
Seriously postgres? There's a primary key - that's an unique constraint!

I don't understand why our integration tests didn't catch this. So I think we need a small simple test that is going to replicate the issue.

It would be helpful to know what version of pq you're running.

When you say rolled back what do you mean? Did you just run 1.15.0 without regenerating the dB? If so you may be able to now drop in 1.15.1 and the unique constraint will be kept.

To that end it would be helpful to see what the current schema of the issue_index table is.

**Comment by zeripath** [comment]
Similarly what happens if you run `gitea doctor recreate-table issue_index` on 1.15.1?

**Comment by RowanTaubitz** [comment]
Just ran into this issue as well, running `gitea doctor recreate-table issue_index` fixed it for us - thanks!

**Comment by zeripath** [comment]
Hmm this implies that xorm is dropping the unique constraint and not adding in the primary key one. It may be that we need to add the (unnecessary) unique constraint back in here and let that be removed in 1.16 when we regenerate the table.

This doubly raises the importance of telling us the schema of the table.

**Comment by zeripath** [comment]
OK I've just looked at this table now I'm actually up and out of bed.

Unfortunately I think all users upgrading from 1.15.0 to 1.15.1 need to run `gitea doctor recreate-table issue_index` - it's just that postgres is the only one telling us this.

Damn. I totally forgot that Xorm would drop the Unique constraint - (I knew it wouldn't create the PK). If I'd just left the unique constraint in this would have been fine.

**Comment by NepPure** [comment]
The same problem

**Comment by zeripath** [comment]
> The same problem

 `gitea doctor recreate-table issue_index`

**Comment by karthanistyr** [comment]
Log of the above command after upgrading again back to 1.15.1:
```
bash-5.1$ gitea doctor recreate-table issue_index
2021/09/03 11:44:46 ...om/urfave/cli/app.go:405:RunAsSubcommand() [I] PING DATABASE postgres
2021/09/03 11:44:46 ...ations/migrations.go:444:func1() [I]
Recreating Table: issue_index for Bean: IssueIndex
2021/09/03 11:44:46 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table user Column max_repo_creation db default is
'-1', struct default is -1
2021/09/03 11:44:46 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table user Column keep_activity_private db default is , struct default is false
2021/09/03 11:44:46 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table user Column keep_activity_private db nullable is true, struct nullable is false
2021/09/03 11:44:46 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table repository Column original_service_type db default is 0, struct default is
2021/09/03 11:44:47 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table comment has column old_assignee_id but struct has not related field
2021/09/03 11:44:47 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table webhook Column is_system_webhook db default
is false, struct default is
2021/09/03 11:44:47 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table webhook Column is_system_webhook db nullable is false, struct nullable is true
2021/09/03 11:44:47 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table webhook Column http_method db default is 'POST', struct default is
2021/09/03 11:44:47 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table email_address Column lower_email db nullable is true, struct nullable is false
2021/09/03 11:44:47 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table external_login_user column avatar_url db type is TEXT, struct type is VARCHAR(255)
2021/09/03 11:44:47 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table tracked_time Column time db nullable is true, struct nullable is false
2021/09/03 11:44:47 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table review column type db type is VARCHAR(255),
struct type is INTEGER
2021/09/03 11:44:47 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table project column board_type db type is INTEGER, struct type is BIGINT
2021/09/03 11:44:47 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table project column type db type is INTEGER, struct type is BIGINT
2021/09/03 11:44:47 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table session has column created_unix but struct has not related field
```
Creating a PR after that worked on 1.15.1. Thanks and sorry I couldn't give you the table info sooner, I got to bed eventually :sleepy: 

Re:
> Did you just run 1.15.0 without regenerating the dB?

Yes I had done that, simply rolled the container back to 1.15.0 without any further change. Hopefully I did not damage the schema. There is seldom any activity on this instance apart from me so I'm confident there hasn't been any more attempts to create issues and PRs before now.

**Comment by viceice** [comment]
Are the warnings normal? Should i do anything?
```sh
> gitea doctor recreate-table issue_index
2021/09/03 12:13:05 ...om/urfave/cli/app.go:405:RunAsSubcommand() [I] PING DATABASE postgres
2021/09/03 12:13:05 ...ations/migrations.go:444:func1() [I] Recreating Table: issue_index for Bean: IssueIndex
2021/09/03 12:13:06 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table user Column max_repo_creation db default is '-1', struct default is -1
2021/09/03 12:13:07 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table email_address Column lower_email db nullable is true, struct nullable is false
2021/09/03 12:13:07 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table external_login_user column avatar_url db type is TEXT, struct type is VARCHAR(255)
2021/09/03 12:13:08 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table project column board_type db type is INTEGER, struct type is BIGINT
2021/09/03 12:13:08 ...om/urfave/cli/app.go:405:RunAsSubcommand() [W] Table project column type db type is INTEGER, struct type is BIGINT
```

Using this as backend: `PostgreSQL 11.13 on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit`
From Helm Chart: [bitnami postgresql](https://artifacthub.io/packages/helm/bitnami/postgresql)
Image currently pinned to

```yaml
image:
  registry: gcr.io
  repository: bitnami-containers/postgresql
  tag: 11.13.0-debian-10-r7
```

**Comment by zeripath** [comment]
These are fine and can be ignored.

**Comment by kenji21** [comment]
> > The same problem
> 
> `gitea doctor recreate-table issue_index`

same issue with a mysql (and same fix)

**Comment by zeripath** [comment]
Closing as fixed by #16938
